src/ipc/capnp/common-types.h

```diff
@@ -103,6 +139,39 @@ decltype(auto) CustomReadField(TypeList<UniValue>, Priority<1>, InvokeContext& i
         value.read(std::string_view{data.begin(), data.size()});
     });
 }
+
+//! Generic ::capnp::Data field builder for any class that a Span can be
+//! constructed from, particularly BaseHash and base_blob classes and
+//! subclasses. It's also used to serialize vector<char> set elements
+//! in GCSFilter::ElementSet and CBlockTemplate::vchCoinbaseCommitment.
+//!
+//! There is currently no corresponding ::capnp::Data CustomReadField function
+//! that works using Spans, because the bitcoin classes in the codebase like
+//! BaseHash and blob_blob that can converted /to/ Span don't currently have
+//! Span constructors that allow them to be constructed /from/ Span. If they
+//! did, it would simplify things. For example, a generic CustomReadField
+//! function could be written that would allow dropping specialized
+//! CustomReadField functions for types like PKHash.
+//!
+//! For the LocalType = vector<char> case, it's also not necessary to
+//! have ::capnp::Data CustomReadField function corresponding to this
+//! CustomBuildField function because ::capnp::Data inherits from
+//! ::capnp::ArrayPtr, and libmultiprocess already provides a generic
+//! CustomReadField function that can read from ::capnp::ArrayPtr into
+//! std::vector.
+template <typename LocalType, typename Value, typename Output>
+void CustomBuildField(
```

Comment:
> The argument list is really difficult to read due to the nature of the arguments. However, I understand that it can be no better than that anyway!

To be fair, I think this could be significantly simpler now that we are using c++20, which supports [concepts](https://en.cppreference.com/w/cpp/language/constraints). In general there should be no reason to use std::enable_if anymore now that concepts are available. There's a lot of older code like this that could be switched to use concepts, and I haven't tried to take that on yet, but this could be a good place to start Thanks for pointing out the issue, and I'll see any there are any simple improvements that can be made here.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1757372497

Removed enable_if so this should be simpler now. Thanks for bringing this up!